tools/comdb2ar/serialise.cpp

```diff
@@ -988,10 +1173,62 @@ void serialise_database(
         serialise_file(FileInfo(FileInfo::LOG_FILE, absfile, dbdir));
 
         long long log_number(lowest_log);
-        for(std::list<FileInfo>::const_iterator
-                it = data_files.begin();
-                it != data_files.end();
-                ++it) {
+
+        // Write the header
+        TarHeader head;
+        head.set_filename(getDTString() + ".data");
+        head.set_attrs(st);
+        head.set_checksum();
+
+        if(writeall(1, head.get().c, sizeof(tar_block_header))
+                != sizeof(tar_block_header)) {
+            std::ostringstream ss;
+            ss << "error writing tar block header: " << std::strerror(errno);
+            throw Error(ss.str());
+        }
+
+        // Iterate through the files, index-aligning with the list of changed pages
+        std::vector<FileInfo>::const_iterator data_it = incr_data_files.begin();
+        std::vector<std::vector<uint32_t>>::const_iterator
+            page_it = page_number_vec.begin();
+
+        ssize_t data_written = 0;
+        // First, serialise any complete log files that are in the .txn
```

Comment:
The order is wrong here: you have to serialize changed pages before logfiles.  So, record all changed pages, then record all logfiles.  Logfiles have to be last or recovery will not work.